### ISSUE TITLE
NativeCall CArrays of CStructs broken

### ISSUE THREAD

**Issue by rakool** [rakudo/rakudo#4022]
NativeCall is advertised to  support C arrays.  CArrays are "_constrained to work with [various C types and]  arrays of structs..."._

**Comment by salortiz** [comment]
The actual NativeCall documentation misleads the fact that a `CArray`, when parameterized with a `CStruct` type `Foo`, it only allocate space for a `Pointer[Foo]`, not a concrete `Foo`. That behavior can't be changed now, a lot of code depends on it.

But let `zef` search for  NativeHelpers::CStruct (included in my [NativeHelpers-Blob]( https://github.com/salortiz/NativeHelpers-Blob) module), it provides a `LinearArray` `role` paremeterizable to any `CStruct` type that allows to work with real C array of structs.

Please see an [example](https://github.com/salortiz/NativeHelpers-Blob/blob/master/t/02-cstruct.t) of it use.

**Comment by rakool** [comment]
@salortiz , thank you for the reference and for your efforts that improved the situation.

That, for years, the erroneous code is in Rakudo and the correct code is elsewhere says what?

I disagree.  It cannot be changed inline, but it can be changed. Maybe re-`used`? A `NativeCall2`  seems a sound, simplistic and fast approach.   
 
Does `NativeCall` fall under `use v6.X`s jurisdiction?  That could help expedite change.

**Comment by jonathanstowe** [comment]
I don't believe that it is "erroneous" or "correct" code in either case, both styles of interface exist in C libraries in the wild.  The "flat array of struct" is really only sugaring over a block of memory that is some multiple of the size of the struct (which is how @salortiz implemented it in the module,) so it wouldn't really need a new version of NativeCall but simply some way of distinguishing between  the flat and pointer arrays and the respective marshalling between raku and C.  Someone just needs to write the code so both can be used with the appropriate annotation.

BTW the same applies to struct arguments to functions.

**Comment by rakool** [comment]
@jonathanstowe, exactly.    What is wrong with the code is that:

- `CArray[aCStruct]`  misnames its function, contradicting its own and `C`'s conventions.                                          
- @salortiz's `LinearArray` functionality is missing.

My thought was that creating a new namespace is not about rewriting `NativeCall` but
about these requirements:
- combining @salortiz's pertinent code or equivalent.
- have a namespace with good naming: `NArray[CStruct]` (and what? `NArray[Pointer[CStruct]]`).   
- have `NativeCall` add in the functionality  with `NArray` and `CArray:adverb`.
- allow users the most flexibility about when to revise or migrate.   
- to surprise inattentive users with no more than a `Could not find TheModule in:`.  
 
Off-hand, I think all those could be met by adding `NArray` into `NativeCall`, but due to its low level nature I'm not sure.  I suggest this requirement list as most helpful to the most peoplem, from devs of all sorts to users of unsupported code.

Thank you.
